# Data Import

In [1]:
import pandas as pd
import numpy as np
from sklearn import metrics
pd.options.mode.chained_assignment = None

main_data = pd.read_csv("C:/Users/Asus/CSE475 project/diabetes.csv")
main_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1


# Data Preprocessing

In [2]:
import statistics

label_array = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome']
for i in range(1, 6):
    label = label_array[i]
    x = []
    
    for j in range(0, len(main_data)):
        if(main_data[label][j] != 0):
            x.append(main_data[label][j])
        
    med = statistics.median(x)
    
    for j in range(0, len(main_data)):
        if(main_data[label][j] == 0):
            main_data[label][j] = med
    
main_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35      125  33.6   
1            1       85             66             29      125  26.6   
2            8      183             64             29      125  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [3]:
x = main_data.iloc[:, :-1]
x.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35      125  33.6   
1            1       85             66             29      125  26.6   
2            8      183             64             29      125  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
2                     0.672   32  
3                     0.167   21  
4                     2.288   33

In [4]:
y = main_data.iloc[:, 8:]
y.head()

Outcome
0        1
1        0
2        1
3        0
4        1

# Splitting Train and Test Data

In [5]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)
y_test = y_test.to_numpy().ravel()
y_train = y_train.to_numpy().ravel()

# Decision Tree

In [6]:
from sklearn import tree

decisionTreeClf = tree.DecisionTreeClassifier(random_state=0)
decisionTreeClf.fit(x_train, y_train)
decisionTreeprediction = decisionTreeClf.predict(x_test)
r1 = metrics.classification_report(y_test, decisionTreeprediction)
print(r1)

              precision    recall  f1-score   support

           0       0.86      0.76      0.81        96
           1       0.67      0.79      0.72        58

    accuracy                           0.77       154
   macro avg       0.76      0.78      0.77       154
weighted avg       0.79      0.77      0.78       154



In [7]:
decision_tree_confusion_metrix = metrics.confusion_matrix(y_test, decisionTreeprediction)
pd.DataFrame(decision_tree_confusion_metrix)

0   1
0  73  23
1  12  46

# Random Forest

In [8]:
from sklearn.ensemble import RandomForestClassifier

randomForestClf = RandomForestClassifier(random_state=0)
randomForestClf.fit(x_train, y_train)
randomForestPrediction = randomForestClf.predict(x_test)
r2 = metrics.classification_report(y_test, randomForestPrediction)
print(r2)

              precision    recall  f1-score   support

           0       0.83      0.85      0.84        96
           1       0.75      0.71      0.73        58

    accuracy                           0.80       154
   macro avg       0.79      0.78      0.78       154
weighted avg       0.80      0.80      0.80       154



In [9]:
random_forest_confusion_metrix = metrics.confusion_matrix(y_test, randomForestPrediction)
pd.DataFrame(random_forest_confusion_metrix)

0   1
0  82  14
1  17  41

# K-Nearest Neighbors

In [10]:
from sklearn.neighbors import KNeighborsClassifier

knnClf = KNeighborsClassifier(n_neighbors=7)
knnClf.fit(x_train, y_train)
knnPrediction = knnClf.predict(x_test)
r3 = metrics.classification_report(y_test, knnPrediction)
print(r3)

              precision    recall  f1-score   support

           0       0.81      0.88      0.84        96
           1       0.76      0.66      0.70        58

    accuracy                           0.79       154
   macro avg       0.78      0.77      0.77       154
weighted avg       0.79      0.79      0.79       154



In [11]:
knn_confusion_metrix = metrics.confusion_matrix(y_test, knnPrediction)
pd.DataFrame(knn_confusion_metrix)

0   1
0  84  12
1  20  38

# Support vector machine

In [12]:
from sklearn import svm

svmClf = svm.SVC()
svmClf.fit(x_train, y_train)
svmPrediction = svmClf.predict(x_test)
f4 = metrics.classification_report(y_test, svmPrediction)
print(f4)

              precision    recall  f1-score   support

           0       0.76      0.95      0.85        96
           1       0.86      0.52      0.65        58

    accuracy                           0.79       154
   macro avg       0.81      0.73      0.75       154
weighted avg       0.80      0.79      0.77       154



In [13]:
svm_confusion_metrix = metrics.confusion_matrix(y_test, svmPrediction)
pd.DataFrame(svm_confusion_metrix)

0   1
0  91   5
1  28  30

|  | Precision | Recall | F1-score | Accuracy |
| --- | --- | --- | --- | --- |
| Decision Tree | 0.76 | 0.78 | 0.77 | 0.77 |
| Random Forest | 0.79 | 0.78 | 0.78 | 0.80 |
| K-Nearest Neighbors | 0.78 | 0.77 | 0.77 | 0.79 |
| Support Vector Machine | 0.81 | 0.73 | 0.75 | 0.79 |

|  | True Positive | True Negative | False Positive | False Negative |
| --- | --- | --- | --- | --- |
| Decision Tree | 46 | 73 | 23 | 12 |
| Random Forest | 41 | 82 | 14 | 17 |
| K-Nearest Neighbors | 38 | 84 | 12 | 20 |
| Support Vector Machine | 30 | 91 | 5 | 28 |